### ISSUE TITLE
Ansible connections using warpgate are slow

### ISSUE THREAD

**Issue by ntimo** [warp-tech/warpgate#459]
I just noticed that when I use Ansible with Warpgate the playbook execution time is slower.

## Test setup
I run a playbook against the host where warpgate is installed via systemd.
For the authentication with Ansible I used the password login for warpgate.
For the test without warpgate I used a ssh private key to connect to the target host directly by Ansible.
I then run the Ansible playbook with the time command to measure the time it took to complete.

# Results:
warpgate (ssh with password auth):
```
real    0m44.063s
user    0m3.355s
sys     0m0.567s
```

warpgate (ssh with key auth):
```
real    0m28.975s
user    0m2.655s
sys     0m0.414s
```

direct:
```
real    0m4.923s
user    0m1.212s
sys     0m0.338s
```

**Comment by stappersg** [comment]
Please rerun the test also with _SSH keys_ when going through the `warpgate`.

**Comment by ntimo** [comment]
@stappersg I did rerun the test and it was defiantly faster but not nearly as faster but not really great. 28s seconds vs 4 is still a huge difference. And the playbook I use to test this only contains 2 Ansible template tasks.

**Comment by ntimo** [comment]
I also rerun the test 3 times and the numbers are always about the same

**Comment by stappersg** [comment]
@ntimo  Please relax.   Do know that your reported issue got human attention.

And pretty please dive deeper into this issue. Do more measurement to find out where `warpgate` is spending time. Try to understand that `warpgate` sits between _ansible control machine_ and the _ssh target_, by definition it causes progation delay. There is no added value in expressing _"is still a huge difference"_. If you don't want to pay for the benefits that `warpgate` brings above a _SSH jumphost_, then don't use `warpgate`.

No hard feelings, just an attempt to shift you from _"the delay should be reduced"_ to _"(help to) make it possible to reduce the delay"_.

**Comment by stappersg** [comment]
Only linking to another [`ansible issue`](https://github.com/warp-tech/warpgate/issues/270#issuecomment-1221397200).

**Comment by ntimo** [comment]
@stappersg thank you. I never actually wanted to be rude or unpolite. I am sorry if the way I worded the issue made it feel that way. And thank you for your time. I will do more testing and report back to you.

**Comment by Abraxos** [comment]
Hello,

Not trying to pile on or anything, but I've been experiencing an issue with slowness as well. I've been gathering information on it to file a real, helpful bug report, but I figured I would at least comment that I have been dealing with a slowdown issue with both ansible, borg backups, RSync, and SFTP file transfers in general. I was also waiting for a release where https://github.com/warp-tech/warpgate/issues/270#issuecomment-1304830816 would be fixed.

Should I post my findings on this bug, or should I file a completely separate bug? Also should I bother waiting for a fix to https://github.com/warp-tech/warpgate/issues/270#issuecomment-1304830816 or report as soon as possible?

Thank you in advance.

**Comment by stappersg** [comment]
> ... but I've been experiencing an issue with slowness as well. ... Should I post my findings ...

Providing facts is indeed a good thing.
IMNSHO: Go for a fresh issue with a title like _"Propagation delay in comparison with SSH jumphost"_.

**Comment by ntimo** [comment]
I actually found something out, I tested ansible in combination with mitogen https://mitogen.networkgenomics.com/ansible_detailed.html which only uses 1 persistent ssh connection for the entire playbook run against a given host. Where ansible opens a new connection per task which is probably what makes it so slow. Using mitogen I had really really fast ansible playbook times. But I will post more details later, first need to do more tests. I also want to try to use the ssh ControlMaster feature.

**Comment by Eugeny** [comment]
I've just pushed an update that should eliminate the initial 1s pre-authentication delay for each new SSH connection

**Comment by ntimo** [comment]
Hey @eugene thank you for your time and effort I just updated my warpgate and tested this again with `pipelining = True` enabled I now get these times. I run the exact same playbook 4 times after each other. And as you can see the first run was quite fast while all others after it where slower. But if I wait some time lets say 30 minutes the first run is going to be fast again. Could this be some kind of a caching problem? I find this is kind of strange.
after update (using warpgate):
```
 0:00:06.016 ******
 0:00:18.989 ******
 0:00:19.021 ******
 0:00:18.980 ******
```

When running it directly against the node without warpgate in the middle I get these consitant times:
```
0:00:02.884 ******
0:00:02.504 ******
0:00:02.494 ******
0:00:02.505 ******
```

**Comment by stappersg** [comment]
> ... And as you can see the first run was quite fast while all others after it where basically twice as slow. But if I wait some time lets say 30 minutes the first run is going to be fast again. Could this be some kind of a caching problem? I find this is kind of strange.


Yes, an interesting problem.   Usualy is the first one slow and are the next ones faster (due _cache_ hit), here is first one the fast one and the other ones slower  (struggling with internal data structures??)

Path I would like to travel is the path of a scientific approach, verifiable measurements.

@ntimo Please describe both setups that you are using for your measurements.  With _both_ I mean the _warpgate setup_ and the _reference setup_.  Do know that it is Okay to avoid perfect descriptions, that we will learn along the way. ( Like learning tooling as https://github.com/flamegraph-rs/flamegraph . )

**Comment by ntimo** [comment]
@stappersg I just created a guide how to setup a reproducible test environment in the Hetzner Cloud:
https://github.com/ntimo/warpgate-issue-459

**Comment by stappersg** [comment]
On Sat, Nov 12, 2022 at 08:36:36AM -0800, Timo wrote:
> @stappersg I just created a guide how to setup a reproducible test environment in the Hetzner Cloud:
> https://github.com/ntimo/warpgate-issue-459
> 

Acknowledge on one setup.
(as in:  my first (way too??) quick scan saw 1 setup)


Groeten
Geert Stappers
By email, aiming for direct communication  (communication without Github.com)
-- 
Silence is hard to parse

**Comment by ntimo** [comment]
I just run a test by running the playbook 50 times each direct / using a jumphost / using warpgate.
Here is the result:
![plot](https://raw.githubusercontent.com/ntimo/warpgate-issue-459/task/jumphost-vs-direct-vs-warpgate/results/50-runs/plot.png)

**Comment by stappersg** [comment]
> I just run a test by running the playbook 50 times each direct / using a jumphost / using warpgate.

Nice!


![afbeelding](https://user-images.githubusercontent.com/35434094/201499729-23c56cc2-7839-4c57-9008-a84a4ff6a7a5.png)

**Comment by Abraxos** [comment]
> > ... but I've been experiencing an issue with slowness as well. ... Should I post my findings ...
> 
> Providing facts is indeed a good thing. IMNSHO: Go for a fresh issue with a title like _"Propagation delay in comparison with SSH jumphost"_.

I went to actually collect the data and discovered that speed is not the issue (at least not a primary issue). It seems like the connection is being dropped for some reason. I give details here: https://github.com/warp-tech/warpgate/issues/479 but I believe the issue can be reproduced simply by transferring a file of non-trivial size using `scp` (unless the issue is somehow unique to my setup). Funnily enough, `rsync` works just fine and is exactly as fast between a direct connection, a jumphost, and a Warpgate host.

**Comment by ntimo** [comment]
@Abraxos I was also able to reproduce the broken pipe issue with scp. I also see a few broken pipe messages in the ansible logs when running ansible with `-vvvv`.

**Comment by ntimo** [comment]
I have uploaded the ansible logs here: https://github.com/ntimo/warpgate-issue-459/tree/master/results/ansible_log_broken_pipe

**Comment by Abraxos** [comment]
> @Abraxos I was also able to reproduce the broken pipe issue with scp. I also see a few broken pipe messages in the ansible logs when running ansible with `-vvvv`.

Cool! Thank you for reproducing and confirming. Do you believe that the issues are related? It seems obvious that broken pipes would not be helpful to ansible, but at the same time, they don't appear to happen to every command, but every command is slowed.

**Comment by ntimo** [comment]
@Eugeny after the latest update compared the the previous Warpgate version there is quite a difference in Ansible times:
![plot](https://raw.githubusercontent.com/ntimo/warpgate-issue-459/master/results/50-runs/plot-commit-f6a73c0.png)

**Comment by ntimo** [comment]
Okay I just found something out. Ansible does use 1 connection to warpgate on the first run to run all tasks connection, this is also visable in warpgate since the session has a bunch of recordings. But the sessions is not closed when Ansible finishes the playbook which seems to be intended .... see (https://github.com/ansible/ansible/issues/78227) that's why the first run is so fast. 

The second run where the session at warpgate is still open is then using single sessions for every connection for some stupid reason (but I guess that's just Ansible beeing special).

You can fix the "problem" with Ansible not closing the session by running this as the last task in your playbook.
```
- name: Close connection
  meta: reset_connection
  tags: always

```

Otherwise you have to wait for warpgate to force close the session after a few seconds.

None the less warpgate is still a bit slower then using a normal jumphost, and I am not sure if the aim of warpgate is to be as fast as a "normal" jumphost. I will also research a bit more on how to `reset_connection` maybe do the connection reset by default if it's even possible.

**Comment by stappersg** [comment]
On Tue, Nov 15, 2022 at 11:48:57AM -0800, Timo wrote:
> 
> You can fix the "problem" with Ansible not closing the session by
> running this as the last task in your playbook.
>   - name: Close connection
>     meta: reset_connection
>     tags: always
> 
> Otherwise you have to wait for warpgate
> to force close the session after a few seconds.

Oh, that explains a lot.


Groeten
Geert Stappers
-- 
Silence is hard to parse

**Comment by ntimo** [comment]
> Okay I just found something out. Ansible does use 1 connection to warpgate on the first run to run all tasks connection, this is also visable in warpgate since the session has a bunch of recordings. But the sessions is not closed when Ansible finishes the playbook which seems to be intended .... see ([ansible/ansible#78227](https://github.com/ansible/ansible/issues/78227)) that's why the first run is so fast.

This only happens for the first 10 recordings in 1 warpgate session since the sshd default `MaxSessions` value is 10, after that Ansible uses 1 warpgate session again for everything else. This can be worked arround by chaning the sshd `MaxSession` limit to lets say `1000`  then when running Ansible ansible is going to use 1 warpgate session with 20+ recodings.

**Comment by Eugeny** [comment]
I've just found out that Warpgate didn't correctly close SSH channels - leading to `MaxSessions` getting exhausted - now the whole playbook should run through the same connection

**Comment by ntimo** [comment]
@Eugeny I just tested the fix and it works perfectly, I changed back the MaxSessions back to 10 and let a playbook run and one warpgate sessions was used with more then 10 recordings 👍

**Comment by Eugeny** [comment]
Thanks for checking - what numbers are you getting now in comparison to a normal jumphost?

**Comment by ntimo** [comment]
@Eugeny I just tested it with the test script I had written and the times for direct ssh and warpgate are basically identical.

Just a few samples:
SSH:
```
00:02.887 **
00:02.292 **
00:02.257 **
00:02.181 **
00:02.163 **
00:02.234 **
00:02.226 **
00:02.047 **
```

Warpgate:
```
00:03.016 **
00:02.559 **
00:02.678 **
00:02.517 **
00:02.550 **
00:02.594 **
00:02.567 **
```
![image](https://user-images.githubusercontent.com/6145026/203148202-6d000b35-5ce0-45c5-ba62-2e01bc104ee8.png)

**Comment by Eugeny** [comment]
This is great - thank you so much for testing!

**Comment by ntimo** [comment]
@Eugeny Thanks,
I am currently running one last test where I have Ansible copy a 1GB file to the host using warpgate and measuring the times. And it seems like this is slower compared to jumphosts by about 7 seconds per run, but I guess scp using warpgate is generally slower then using a jumphost. So that's probably not really Ansible related.

Here are some numbers:
direct:
```
00:11.487 **
00:10.906 **
00:10.744 **
00:10.776 **
00:10.553 **
00:10.733 **
00:10.566 **
00:11.575 **
00:11.272 **
00:11.128 **
00:10.878 **
```
jumphost:
```
00:10.798 **
00:11.000 **
00:10.731 **
00:11.461 **
00:11.149 **
00:11.768 **
00:10.937 **
00:11.368 **
00:10.825 **
00:10.886 **
00:10.879 **
00:10.674 **
```
warpgate:
```
00:21.423 **
00:17.470 **
00:16.932 **
00:17.017 **
00:17.649 **
00:17.004 **
00:17.477 **
05:16.055 **
00:15.131 **
00:16.564 **
```
